### PR TITLE
Add --datapath option

### PR DESCRIPTION
**Pull request by Ansa89**
-- @grayhatter edits

- [x] Windows support
- [x] OSX support
- [x] Ignore `--portable`
- [x] Error checking
    - [x] Exit on error
    - [x] Warn on long path lengths
    - [x] It shouldn't create the directory for the user, it should just fail when the directory doesn't exist. Or when the user lack the necessary permissions to interact with the file/directory
- [x] Rebase or merge (after these are fixed)

### CODE REVIEW

**Comment by tsudoko** [comment]
Option formatting is extremely inconsistent, the theme option uses `--opt arg`, while your datapath option uses `--opt=arg`. I think it should conform to the existing way of doing things or support both formats.

**Comment by Ansa89** [comment]
My fault to not search for similar open PR.
Why notsecure/uTox#1010 isn't already merged?

**Comment by Ansa89** [comment]
@tsudoko: now it should work also `--datapath example/dir`.

**Comment by LittleVulpix** [comment]
@Ansa89 No idea why it's not merged. But I am curious how these two end up interacting with each other. Sounds like a recipe for disaster : p

**Comment by Ansa89** [comment]
> I am curious how these two end up interacting with each other
They won't: only one PR will be merged.

**Comment by SmoothDude** [comment]
Why only one ? They aren't conflicting with each other and I think it is nice idea to set up custom path for utox data.

**Comment by Ansa89** [comment]
@SmoothDude: I thought your PR was the ~same as mine, but now I see the difference. Moreover if you say they don't conflict in any way, then the dicision is up to @irungentoo.

**Comment by GrayHatter** [comment]
They might both be merged, or maybe neither. They don't conflict and both do different things.

- [ ] Windows support
- [ ] error checking
  - [ ] it shouldn't create the directory for the user, it should just fail when the directory doesn't exist. Or when the user lack the necessary permissions to interact with the file/directory
- [ ] rebase or merge (after these are fixed)

**Comment by Ansa89** [comment]
@GrayHatter:
- [x] windows support
- [x] osx support
- [x] don't create directories
- [x] rebase

File: cocoa/main.m [hunk 1]
```diff
@@ -228,6 +253,12 @@ int datapath(uint8_t *dest) {
         dest[l++] = '/';
 
         return l;
+    } else if (user_defined_datapath) {
```
Comment (GrayHatter) [inline cocoa/main.m]: is there a reason you think portable should override --datapath?

**Comment by GrayHatter** [comment]
@Ansa89 this looks good, if you can change the order in my last comment, or give a reason you want --portable to override --datapath It'll be good to go

**Comment by Ansa89** [comment]
@GrayHatter: what should be the correct behavior when both `--portable` and `--datapath` options are given at the same time?

**Comment by GrayHatter** [comment]
@Ansa89 if --datapath is set, and it resolves to a valid directory that exsts and in readable and writable, utox should ignore the --.portable switch. theres' two votes for datapath (--datapath & the datapath itself) and only one for portable....

that or utox should refuse to start becasue they options are conflicting

**Comment by LittleVulpix** [comment]
I think --portable is basically implicitly 'on' when you specify --datapath, because all you are doing is specifying portable mode - with the actual directory.

**Comment by Ansa89** [comment]
@GrayHatter: now `--portable` is ignored if a valid `--datapath` is specified.
@LittleVulpix: the main difference is that `--portable` is able to create its directory, while `--datapath` rely on an already existing path.

File: cocoa/main.m [hunk 2]
```diff
@@ -97,6 +101,27 @@ int parse_argv(int argc, char const *argv[]) {
...
+                        debug("Using \"%s\" as data path\n");
+                    }
+                } else if((strlen(argv[i]) == 10) && (argc > i+1)) {
+                    i += 1;
+                    strncpy(user_datapath, argv[i], PATH_MAX-1);
+                    user_datapath[PATH_MAX-1] = '\0';
+                    if ((strlen(user_datapath) > 0) && (access(user_datapath, R_OK|W_OK)) == 0) {
+                        user_defined_datapath = 1;
+                        debug("Using \"%s\" as data path\n");
+                    }
```
Comment (GrayHatter) [inline cocoa/main.m]: else `exit(6);` we can make 6 the new error code for filesystem errors

File: cocoa/main.m [hunk 3]
```diff
@@ -97,6 +101,27 @@ int parse_argv(int argc, char const *argv[]) {
...
+                } else {
+                    debug("Ignoring \"--datapath\" argument\n");
```
Comment (GrayHatter) [inline cocoa/main.m]: `exit(6);`

**Comment by GrayHatter** [comment]
If path-given-by-user >450 char, warn about path length and data loss

**Comment by GrayHatter** [comment]
what happens when the last char is or isn't a / (or a \\ on windows)

**Comment by GrayHatter** [comment]
see OP

**Comment by Ansa89** [comment]
> If path-given-by-user >450 char, warn about path length and data loss

Why 450 chars?
Isn't enough `debug("Using \"%s\" as data path\n", user_datapath);`?

> what happens when the last char is or isn't a / (or a \ on windows)

In both cases an additional "/" or "\" is added when the function `int datapath(uint8_t *dest)` is called.

**Comment by GrayHatter** [comment]
I picked 450 out of the air, but UTOX_FILE_NAME_LENGTH is 1024, when I wrote it was was only 512, so I was worried about overruns with long filenames. This shouldn't be a problem here.

**Comment by GrayHatter** [comment]
@Ansa89 looks good man thank you! 

Once @irungentoo takes a look he'll merge it in.

**Pull request closed without merging**